### ISSUE TITLE
[Wasm Proposal] Target the Wep API instead of using  .NET runtime on the client

### ISSUE THREAD

**Issue by VBAndCs** [mono/mono#19650]
I have an idea for mono wasm to minimize the .net dlls downloaded to the client. Instead of using .NET classes on client, we can use the Web API supported by all browsers to do the same stuff JS does.
To do so, we need two things:
1. We need to build a wrapper API in .net, so Blazor code can use it to target the web API.
This is a good example of such a wrapper: https://github.com/nikhilk/scriptsharp/tree/cc/src/Libraries/Web
It is part pf an old project called "Script#" that compiles C# code to JS. The classes in this wrapper are empty signatures (methods and properties return nulls, and voids have empty bodies)

2. Mono wasm should compile the wrapper code to wasm code that targets web API.
Wasm team told me they are working on [interface-types](https://github.com/WebAssembly/interface-types/blob/master/proposals/interface-types/Explainer.md)
I think this will allow you to contact with web API easily.

If this happens, it will:
1. restrict developers from referencing unnecessary dlls.
2. save the build time used to strip out unused functionality from dlls.
3. Save client's space and server's bandwidth/traffic.
3. Always use AOT compiled framework, because the wrapper framework is pre-built for .net Core and app-independent.

**Comment by VBAndCs** [comment]
@BrzVlad @marek-safar @vargaz @lewing
Any response here please?

**Comment by marek-safar** [comment]
We are trying to make .NET Wasm as compatible with .NET as possible and that limits what underlying API we can use. However, this is still an exciting scenario but it should be implemented separately for users who prefer size/performance over compatibility.

**Comment by VBAndCs** [comment]
Hi @marek-safar
About compatibility:
Not only ASP.NET developers are familiar with web API, but also .net desktop developers already use WebControl and mshtml liberty (which is the reason I thought of this idea).
In fact we can use a version mshtml.dll written for .net 5 (It is a COM component now, and can be used in .NET core but on windows only as I think).
This leaves the WebAPI basic types like numeric and strings… etc. We can use this API as a part of the mshtml library, or create a wasm library as limited version of the .net BCL that can  be easily translated to the corresponding webAPI.
By the way, in addition to blazor components, this scenario can allow using a wpf-like design:
1. A file with some html elements, some with names and events. For example:
```
<form name="form1" method="post">
   <button obclick="button_click">
</form>
```

2. a Code behind file that is derived from BazorModel class and contains the button_click handler:
```
partial class Test: BlazorModel
{
   [RunAtClient]
   void button_click(HtmlElement e)
   {
      var txt = New HtmlTextArea();
      form1.insertBefore(txt, e);
   }
}
```

Note that I used the `form1` name as a var. It should be defined in an auto-generated partial class like this:
```
partial class Test 
{
   HtmlDocument Document = GetTheDoc();
   HtmlForm form1 = Document.GetElementByName("form1");
}
```

Benefits of this:
1. Manipulate the DOM as you want.
2. Easily convert existing JS code to Blazor.
3. Allow using Blazor without using Razor code. This is useful for languages other than C#, as they don't support razor.
4. We can create a XAML library for wasm, so desktop developers can easily move to web.
5. We can create Razor Pages that run at server but have some elements with events with handlers marked as `[RunAtClient]`, so they are compiled to wasm and work as AJAX. We don't have to add Blazor components to Razor Pages.
6. We can build blazor libraries independent of html code, so we can use with any Razor page.

I think there can be more applications for this idea. But above all, it will minimize the download size to the client, and save the server bandwidth and traffic.
Thanks.

**Comment by marek-safar** [comment]
I think what you are describing is what I'd love to see happen and we use platform APIs binding term for such APIs. It does not have to go into the framework but can work on top of the framework because the framework is flexible enough to not bring the unused code if you decide to use the only browser-specific APIs.

I'm wondering if there is enough community interest for us to do the scaffolding and general concepts and ask the community to work on implementing the binding APIs.

**Comment by VBAndCs** [comment]
I am willing to participate. I know little about wasm, but I can work on the. Net core side.

**Comment by jaykrell** [comment]
It's a probably of .NET being language or library or both.
It depends who/when you ask.
For example there is a demo of using C# to write a UEFI app.

**Comment by VBAndCs** [comment]
This has nothing to do with the language. Mono translates il to wasm, so, this has nothing to do with c#. It should be a framework liberary, so it can be used by all languages (if there is bo proplem with the il they emit) anf from any. Ner platform such as blazor, razor pages, xamarin and even wpf if it wants to run in the browser. This is my opinion.

**Comment by jaykrell** [comment]
I understand the IL vs. C# factoring, really, but C# language is a big reason people use .NET.
And F#. And the libraries. Anyway, same thing, it is question of what do people want, where do they see value, and if they see the parts that make up the whole.

**Comment by VBAndCs** [comment]
@marek-safar 
Suppose we have the C# library (I can write it), how can I compile the methods targeting it to wasm by mono, to have a pure minimum size wasm that is independent on .NET CRT? Is this available today, or need work at your side?
I want to start working as soon as possible, as I currently have some free time.

**Comment by VBAndCs** [comment]
I referenced the MSHTML.dll to a .NET Core app, and I decompiled the auto-generated Interop.MSHTML.dll file to get a C# project containing the MSHTML wrapper types. It contains more 4540 cs files!
It needs some work to:
1. get rid of unneeded types for wasm.
2. get rid of the ComImport attributes and alike.
3. Changes the names to be shorter and more readable.
4. add some alias attribute to fix the type names casing differences between c# and WEBAPI. This will also help with point 3.
5. the 4.5 k files resides in one folder! We need to organize them in subfolders.

This is not a trivial work, but it is a minimum effort comparing to creating all these files from scratch!

I can work on this, but it will be faster if we divide the work on many contributors. (I can create a repo if we are OK with this idea and ready to go). This leaves us with the wasm type system and alike, which I hope someone help to write.

**Comment by VBAndCs** [comment]
@marek-safar
Luckily, I found a 2-years old project doing it:
https://github.com/kjpou1/wasm-dom
I think we should fork it and add the wasm type system.

**Comment by VBAndCs** [comment]
@kjpou1
I found your [excellent work](https://github.com/kjpou1/wasm-dom), and I hope you help implement the ideas mentioned in this issue and in this [comment](https://github.com/mono/mono/issues/19650#issuecomment-621244830)

**Comment by marek-safar** [comment]
I agree Kenneth's work is a great starting point to work on more complete version of DOM binding

**Comment by VBAndCs** [comment]
@marek-safar
Indeed. But I stumbled in the https://github.com/mono/mono/issues/19698 issue, and can't do anything before getting the app working. I have no experience about mono nor wasm, and  need a push here or I could stuck forever.

**Comment by VBAndCs** [comment]
@marek-safar
I think we can build on the TypeScript DOM API:
https://github.com/microsoft/TypeScript/blob/master/src/lib/dom.generated.d.ts
It is a one huge file, that needs to be converted to C#, but I can't find a suitable converter to avoid doing this manually! Any ideas?

**Comment by marek-safar** [comment]
That file is generated so it's expected not to be easily readable. I think using TS as the input makes sense but you'll need to write custom convertor.

There are some existing ones but I think there will be a need for particular customizations.

- https://github.com/mono/TsToCSharp
- https://github.com/robin-han/language-converter

**Comment by VBAndCs** [comment]
@marek-safar 
TypeScript uses this tool to generate the files from IDL JSON files, used by Edge:
https://github.com/microsoft/TSJS-lib-generator/
The inputfiles directory contains the IDL files.
Maybe we need to have a similar tool to generate the C# DOM API, so, we can update it and keep it synchronous with Edge and TS.

**Comment by VBAndCs** [comment]
A good news: we can use [MIDL](https://docs.microsoft.com/en-us/windows/win32/com/midl-compiler) to convert IDL to tlb which can be used to generate a wrapper dll. There maybe some things to do to make this work, and I hope you help if you have experience in this aria.

**Comment by MarcosCobena** [comment]
Hey @VBAndCs,

We've already tried MIDL for wrapping WebGL IDL files with no success. After some attempts we ended up building [a custom transpiler](https://github.com/WaveEngine/WebGL.NET/tree/master/src/WebIDLToCSharp) to turn IDL into C#. However, if you look for performance, this' not the better approach nowadays: it's way better to rely on "native" libraries consumed through Emscripten, as Uno guys do.

@kjpou1's excellent work with wasm-dom is definitelly something I'd really like to easily install through NuGet, for instance.

My 2 cents :-)

**Comment by cveld** [comment]
My question #19835 got linked to this issue. @kjpou1's wasm-dom project looks like what rust's wasm-bindgen is doing as well? I wonder, is it already possible to directly access DOM objects with the current mono wasm binary? What is the mininum change required? Is it possible to take this part from @kjpou1's work? What are the main challenges?
